# Segmenting and Clustering Neighborhoods in Toronto 

In this assignment, you will be required to explore, segment, and cluster the neighborhoods in the city of Toronto. However, unlike New York, the neighborhood data is not readily available on the internet. What is interesting about the field of data science is that each project can be challenging in its unique way, so you need to learn to be agile and refine the skill to learn new libraries and tools quickly depending on the project.

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. You will be required to scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas dataframe so that it is in a structured format like the New York dataset.

Once the data is in a structured format, you can replicate the analysis that we did to the New York City dataset to explore and cluster the neighborhoods in the city of Toronto.

## There's only one notebook. Please scroll down to review the 3 questions

# QUESTION 1

### 1. For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.



### 2. Start by creating a new Notebook for this assignment.


**Import Libraries**

Basic Libraries

In [41]:
import numpy as np 
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json 

Geo location and Json File 

In [42]:
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim 
import requests 
from pandas.io.json import json_normalize 

Matplotlib for Plotting

In [43]:
import matplotlib.cm as cm
import matplotlib.colors as colors

Sklearn k-means / clustering

In [44]:
from sklearn.cluster import KMeans
from sklearn.datasets.samples_generator import make_blobs

Folium and BeautifulSoup

In [ ]:
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium 
from bs4 import BeautifulSoup
import lxml

### 2A. Get the Data

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe

Note: I had to use a former link of wikipedia because the new one had a different structure
https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=945633050.

Download and Sort it

In [45]:
r = requests.get('https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=945633050.')
soup = BeautifulSoup(r.text, 'html.parser')
table=soup.find('table', attrs={'class':'wikitable sortable'})

Get headers, find rows and skip first one

In [46]:
headers=table.findAll('th')
for i, head in enumerate(headers): headers[i]=str(headers[i]).replace("<th>","").replace("</th>","").replace("\n","")

rows=table.findAll('tr')
rows=rows[1:len(rows)]

### 3. To create the dataframe:

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood.

Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.


More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

### 4. Submit a link to your Notebook on your Github repository. (10 marks)

Note: There are different website scraping libraries and packages in Python. For scraping the above table, you can simply use pandas to read the table into a pandas dataframe.

Another way, which would help to learn for more complicated cases of web scraping is using the BeautifulSoup package. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/

The package is so popular that there is a plethora of tutorials and examples on how to use it. Here is a very good Youtube video on how to use the BeautifulSoup package: https://www.youtube.com/watch?v=ng2o98k983k

Use pandas, or the BeautifulSoup package, or any other way you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe.

**Create the Dataframe**

In [60]:
for i, row in enumerate(rows): rows[i] = str(rows[i]).replace("\n</td></tr>","").replace("<tr>\n<td>","")

df=pd.DataFrame(rows)
df.head()

0
0  M1A</td>\n<td>Not assigned</td>\n<td>Not assigned
1  M2A</td>\n<td>Not assigned</td>\n<td>Not assigned
2  M3A</td>\n<td><a href="/wiki/North_York" title...
3  M4A</td>\n<td><a href="/wiki/North_York" title...
4  M5A</td>\n<td><a href="/wiki/Downtown_Toronto"...

In [61]:
# Expand rows with double lines and drop the old one:
df[headers] = df[0].str.split("</td>\n<td>", n = 2, expand = True) 
df.drop(columns=[0],inplace=True)
df.head()

Postcode                                            Borough  \
0      M1A                                       Not assigned   
1      M2A                                       Not assigned   
2      M3A  <a href="/wiki/North_York" title="North York">...   
3      M4A  <a href="/wiki/North_York" title="North York">...   
4      M5A  <a href="/wiki/Downtown_Toronto" title="Downto...   

                                       Neighbourhood  
0                                       Not assigned  
1                                       Not assigned  
2  <a href="/wiki/Parkwoods" title="Parkwoods">Pa...  
3  <a href="/wiki/Victoria_Village" title="Victor...  
4  <a href="/wiki/Regent_Park" title="Regent Park...

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough

In [62]:
# Ignore cells with a borough that is Not assigned.
df = df.drop(df[(df.Borough == "Not assigned")].index)

# Give "Not assigned" Neighborhoods same name as Borough:
df.Neighbourhood.replace("Not assigned", df.Borough, inplace=True)
df.Neighbourhood.fillna(df.Borough, inplace=True)

# drop duplicate rows:
df=df.drop_duplicates()
df.head()

Postcode                                            Borough  \
2      M3A  <a href="/wiki/North_York" title="North York">...   
3      M4A  <a href="/wiki/North_York" title="North York">...   
4      M5A  <a href="/wiki/Downtown_Toronto" title="Downto...   
5      M6A  <a href="/wiki/North_York" title="North York">...   
6      M6A  <a href="/wiki/North_York" title="North York">...   

                                       Neighbourhood  
2  <a href="/wiki/Parkwoods" title="Parkwoods">Pa...  
3  <a href="/wiki/Victoria_Village" title="Victor...  
4  <a href="/wiki/Regent_Park" title="Regent Park...  
5  <a href="/wiki/Lawrence_Heights" title="Lawren...  
6  <a href="/wiki/Lawrence_Manor" title="Lawrence...

In [63]:
# Extract titles 
df.update(
    df.Neighbourhood.loc[
        lambda x: x.str.contains('title')
    ].str.extract('title=\"([^\"]*)',expand=False))

df.update(
    df.Borough.loc[
        lambda x: x.str.contains('title')
    ].str.extract('title=\"([^\"]*)',expand=False))
df.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto       Regent Park
5      M6A        North York  Lawrence Heights
6      M6A        North York    Lawrence Manor

In [65]:
# Delete Toronto:
df.update(
    df.Neighbourhood.loc[
        lambda x: x.str.contains('Toronto')
    ].str.replace(", Toronto",""))
df.update(
    df.Neighbourhood.loc[
        lambda x: x.str.contains('Toronto')
    ].str.replace("\(Toronto\)",""))
df.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto       Regent Park
5      M6A        North York  Lawrence Heights
6      M6A        North York    Lawrence Manor

In [66]:
# combine neighborhoods with the same post code
df2 = pd.DataFrame({'Postcode':df.Postcode.unique()})
df2['Borough']=pd.DataFrame(list(set(df['Borough'].loc[df['Postcode'] == x['Postcode']])) for i, x in df2.iterrows())
df2['Neighborhood']=pd.Series(list(set(df['Neighbourhood'].loc[df['Postcode'] == x['Postcode']])) for i, x in df2.iterrows())
df2['Neighborhood']=df2['Neighborhood'].apply(lambda x: ', '.join(x))
df2.dtypes

df2.head()

Postcode           Borough                      Neighborhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto                       Regent Park
3      M6A        North York  Lawrence Heights, Lawrence Manor
4      M7A  Downtown Toronto                     Queen's Park

## In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [67]:
df2.shape


(103, 3)

**The number of unique postal codes is 103**

# QUESTION 2

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates. So, in order to make sure that you get the coordinates for all of our neighborhoods, you can run a while loop for each postal code. Taking postal code M5G as an example, your code would look something like this:

Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

Important Note: There is a limit on how many times you can call geocoder.google function. It is 2500 times per day. This should be way more than enough for you to get acquainted with the package and to use it to get the geographical coordinates of the neighborhoods in the Toronto.

Once you are able to create the above dataframe, submit a link to the new Notebook on your Github repository. (2 marks)

**Add Geo-spatial data**

In [107]:
dfll= pd.read_csv("http://cocl.us/Geospatial_data")
dfll.rename(columns={'Postal Code':'Postcode'}, inplace=True)
dfll.set_index("Postcode")
df2.set_index("Postcode")

toronto_data=pd.merge(df2, dfll)
toronto_data.head()

Postcode           Borough                      Neighborhood   Latitude  \
0      M3A        North York                         Parkwoods  43.753259   
1      M4A        North York                  Victoria Village  43.725882   
2      M5A  Downtown Toronto                       Regent Park  43.654260   
3      M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4      M7A  Downtown Toronto                     Queen's Park   43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

# QUESTION 3

Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you.

Just make sure:

1. to add enough Markdown cells to explain what you decided to do and to report any observations you make.
2. to generate maps to visualize your neighborhoods and how they cluster together.

Once you are happy with your analysis, submit a link to the new Notebook on your Github repository. (3 marks)

In [108]:
# Get the latitude and Longitude for Toronto
address = 'Toronto, ON, Canada'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The Coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The Coordinates of Toronto are 43.653963, -79.387207.


**Create a map of Toronto with neighborhoods marked**

In [109]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)


for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

**Use the Foursquare API to explore the neighborhoods and segment them**

In [128]:
# The code was removed by Watson Studio for sharing.

**Explore the neighborhoods with their most frequent venues**

In [136]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

**Create a new dataframe: Toronto Venues**

In [137]:

toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )


Parkwoods
Victoria Village
Regent Park
Lawrence Heights, Lawrence Manor
Queen's Park 
Islington Avenue
Rouge, Malvern
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
Princess Gardens, Martin Grove, Cloverdale, West Deane Park, Islington
Highland Creek , Rouge Hill, Port Union
Don Mills South, Flemingdon Park
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Markland Wood, Old Burnhamthorpe
Morningside, Guildwood, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
King, Adelaide, Richmond
Dufferin, Dovercourt Village
Scarborough Village
Oriole, Fairview, Henry Farm
Northwood Park, York University
East Toronto
Harbourfront East, Toronto Islands, Union Station 
Little Portugal, Trinity–Bellwoods
Ionview, East Birchmount Park, Kennedy Park
Bayview Village
Downsview East, CFB Toronto
The Da

**Check the new dataframe**

In [138]:
print(toronto_venues.shape)
toronto_venues.head()

(4942, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Parkwoods              43.753259              -79.329656   
1    Parkwoods              43.753259              -79.329656   
2    Parkwoods              43.753259              -79.329656   
3    Parkwoods              43.753259              -79.329656   
4    Parkwoods              43.753259              -79.329656   

               Venue  Venue Latitude  Venue Longitude        Venue Category  
0    Allwyn's Bakery       43.759840       -79.324719  Caribbean Restaurant  
1    Brookbanks Park       43.751976       -79.332140                  Park  
2        Tim Hortons       43.760668       -79.326368                  Café  
3                A&W       43.760643       -79.326865  Fast Food Restaurant  
4  Bruno's valu-mart       43.746143       -79.324630         Grocery Store

**Venues returned for each neighborhood**

In [139]:
toronto_venues.groupby('Neighborhood').count()
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Parkwoods              43.753259              -79.329656   
1    Parkwoods              43.753259              -79.329656   
2    Parkwoods              43.753259              -79.329656   
3    Parkwoods              43.753259              -79.329656   
4    Parkwoods              43.753259              -79.329656   

               Venue  Venue Latitude  Venue Longitude        Venue Category  
0    Allwyn's Bakery       43.759840       -79.324719  Caribbean Restaurant  
1    Brookbanks Park       43.751976       -79.332140                  Park  
2        Tim Hortons       43.760668       -79.326368                  Café  
3                A&W       43.760643       -79.326865  Fast Food Restaurant  
4  Bruno's valu-mart       43.746143       -79.324630         Grocery Store

**Dataframe with One Hot Encoding**

In [141]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Zoo  Accessories Store  Afghan Restaurant  Airport  Airport Lounge  \
0    0                  0                  0        0               0   
1    0                  0                  0        0               0   
2    0                  0                  0        0               0   
3    0                  0                  0        0               0   
4    0                  0                  0        0               0   

   American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
0                    0             0               0             0         0   
1                    0             0               0             0         0   
2                    0             0               0             0         0   
3                    0             0               0             0         0   
4                    0             0               0             0         0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0            0                    0                 0                   0   
1            0                    0                 0                   0   
2            0                    0                 0                   0   
3            0                    0                 0                   0   
4            0                    0                 0                   0   

   Auto Dealership  Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  \
0                0            0              0                0          0   
1                0            0              0                0          0   
2                0            0              0                0          0   
3                0            0              0                0          0   
4                0            0              0                0          0   

   Baby Store  Badminton Court  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0           0                0           0       0     0    0               0   
1           0                0           0       0     0    0               0   
2           0                0           0       0     0    0               0   
3           0                0           0       0     0    0               0   
4           0                0           0       0     0    0               0   

   Baseball Stadium  Basketball Stadium  Beach  Beach Bar  Beer Bar  \
0                 0                   0      0          0         0   
1                 0                   0      0          0         0   
2                 0                   0      0          0         0   
3                 0                   0      0          0         0   
4                 0                   0      0          0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Bookstore  Boutique  \
0           0                   0          0       0          0         0   
1           0                   0          0       0          0         0   
2           0                   0          0       0          0         0   
3           0                   0          0       0          0         0   
4           0                   0          0       0          0         0   

   Bowling Alley  Boxing Gym  Brazilian Restaurant  Breakfast Spot  Brewery  \
0              0           0                     0               0        0   
1              0           0                     0               0        0   
2              0           0                     0               0        0   
3              0           0                     0               0        0   
4              0           0                     0               0        0   

   Bridal Shop  Bridge  Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  \
0            0       0                0       0             0              0   
1            0       0                0       0             0              0   
2            0       0                0       0            

**Group rows by neighborhood according to their mean**

In [160]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()

toronto_grouped.head()

Neighborhood  Zoo  Accessories Store  \
0                                        Agincourt  0.0               0.00   
1  Bathurst Manor, Wilson Heights, Downsview North  0.0               0.00   
2                                  Bayview Village  0.0               0.00   
3                                      Berczy Park  0.0               0.00   
4     Brockton, Parkdale Village, Exhibition Place  0.0               0.01   

   Afghan Restaurant  Airport  Airport Lounge  American Restaurant  \
0                0.0      0.0             0.0             0.021739   
1                0.0      0.0             0.0             0.000000   
2                0.0      0.0             0.0             0.000000   
3                0.0      0.0             0.0             0.010000   
4                0.0      0.0             0.0             0.010000   

   Amphitheater  Animal Shelter  Antique Shop  Aquarium  Art Gallery  \
0           0.0             0.0           0.0       0.0         0.00   
1           0.0             0.0           0.0       0.0         0.00   
2           0.0             0.0           0.0       0.0         0.00   
3           0.0             0.0           0.0       0.0         0.01   
4           0.0             0.0           0.0       0.0         0.01   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                 0.00          0.021739                0.00              0.0   
1                 0.00          0.000000                0.00              0.0   
2                 0.00          0.000000                0.00              0.0   
3                 0.00          0.000000                0.00              0.0   
4                 0.02          0.000000                0.01              0.0   

   Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  \
0          0.0            0.0              0.0       0.00         0.0   
1          0.0            0.0              0.0       0.00         0.0   
2          0.0            0.0              0.0       0.00         0.0   
3          0.0            0.0              0.0       0.02         0.0   
4          0.0            0.0              0.0       0.00         0.0   

   Badminton Court  Bagel Shop    Bakery      Bank   Bar  Baseball Field  \
0              0.0        0.00  0.043478  0.021739  0.00             0.0   
1              0.0        0.00  0.000000  0.068966  0.00             0.0   
2              0.0        0.00  0.000000  0.133333  0.00             0.0   
3              0.0        0.01  0.030000  0.000000  0.00             0.0   
4              0.0        0.00  0.040000  0.000000  0.05             0.0   

   Baseball Stadium  Basketball Stadium  Beach  Beach Bar  Beer Bar  \
0               0.0                0.00   0.00       0.00      0.00   
1               0.0                0.00   0.00       0.00      0.00   
2               0.0                0.00   0.00       0.00      0.00   
3               0.0                0.01   0.01       0.00      0.04   
4               0.0                0.00   0.00       0.01      0.01   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Bookstore  Boutique  \
0         0.0                 0.0        0.0    0.00        0.0       0.0   
1         0.0                 0.0        0.0    0.00        0.0       0.0   
2         0.0                 0.0        0.0    0.00        0.0       0.0   
3         0.0                 0.0        0.0    0.01        0.0       0.0   
4         0.0                 0.0        0.0    0.00        0.0       0.0   

   Bowling Alley  Boxing Gym  Brazilian Restaurant  Breakfast Spot  Brewery  \
0            0.0         0.0                   0.0        0.021739     0.00   
1            0.0         0.0                   0.0        0.000000     0.00   
2            0.0         0.0                   0.0        0.000000     0.00   
3            0.0         0.0                   0.0        0.020000     0.00   
4            0.0         0.0                   

**Get each neighborhood with its top 10 most frequent venues**

In [145]:
num_top_venues = 10

for hood in toronto_grouped['Neighborhood']:
    
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    

**Transform into a dataframe of venues sorted in descending order**

In [152]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

**Create the new dataframe and display the top 10 venues for each neighborhood**

In [153]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe of venues sorted
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt    Chinese Restaurant   
1  Bathurst Manor, Wilson Heights, Downsview North     Convenience Store   
2                                  Bayview Village                  Bank   
3                                      Berczy Park           Coffee Shop   
4     Brockton, Parkdale Village, Exhibition Place                  Café   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Shopping Mall                Bakery           Pizza Place   
1                  Bank           Coffee Shop      Sushi Restaurant   
2           Gas Station   Japanese Restaurant         Shopping Mall   
3                 Hotel                  Café            Restaurant   
4           Coffee Shop            Restaurant                   Bar   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Coffee Shop        Sandwich Place  Caribbean Restaurant   
1                   Diner           Gas Station               Dog Run   
2            Skating Rink          Intersection         Grocery Store   
3     Japanese Restaurant              Beer Bar                Bakery   
4  Furniture / Home Store                Bakery             Gift Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Pool           Supermarket  Sri Lankan Restaurant  
1    Frozen Yogurt Shop   Fried Chicken Joint            Supermarket  
2    Chinese Restaurant            Restaurant                   Café  
3             Gastropub    Italian Restaurant                   Park  
4    Tibetan Restaurant        Soccer Stadium            Supermarket

### Cluster Neighborhoods

**Run k-means to cluster the neighborhood into 7 clusters**

In [154]:
kclusters = 7

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

kmeans.labels_[0:10]

array([0, 1, 1, 0, 0, 1, 1, 1, 0, 0], dtype=int32)

**Create a new dataframe that merge Neighborhhoods with its top 10 venues**

In [155]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged.head()

Postcode           Borough                      Neighborhood   Latitude  \
0      M3A        North York                         Parkwoods  43.753259   
1      M4A        North York                  Victoria Village  43.725882   
2      M5A  Downtown Toronto                       Regent Park  43.654260   
3      M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4      M7A  Downtown Toronto                     Queen's Park   43.662301   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.329656             1.0                  Park     Convenience Store   
1 -79.315572             0.0           Coffee Shop          Hockey Arena   
2 -79.360636             0.0           Coffee Shop               Theater   
3 -79.464763             0.0            Restaurant           Coffee Shop   
4 -79.389494             0.0           Coffee Shop                  Park   

  3rd Most Common Venue   4th Most Common Venue  5th Most Common Venue  \
0              Bus Stop                Pharmacy          Shopping Mall   
1  Gym / Fitness Center                    Park                   Café   
2                   Pub                   Diner                   Park   
3  Fast Food Restaurant  Furniture / Home Store  Vietnamese Restaurant   
4             Gastropub      Italian Restaurant              Bookstore   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Pizza Place    Chinese Restaurant  Caribbean Restaurant   
1           Men's Store            Boxing Gym                Lounge   
2                  Café    Italian Restaurant            Restaurant   
3        Clothing Store         Women's Store      Sushi Restaurant   
4           Men's Store           Pizza Place       Thai Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0       Laundry Service            Supermarket  
1           Golf Course          Grocery Store  
2        Breakfast Spot                 Bakery  
3   Fried Chicken Joint           Dessert Shop  
4      Ramen Restaurant    Japanese Restaurant

## Create a map of Toronto with the Clusters

In [ ]:
mapofclusters = folium.Map(location=[latitude, longitude], zoom_start=10)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

**Add markers to the map**

In [158]:
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        fill=True,
        fill_opacity=0.6).add_to(mapofclusters)
       
mapofclusters